core/services/ocr2/plugins/ocr2keeper/evm/registry.go

```diff
@@ -0,0 +1,784 @@
+	if err != nil {
+		return nil, err
+	}
+
+	state, err := r.registry.GetState(opts)
+	if err != nil {
+		n := "latest"
+		if opts.BlockNumber != nil {
+			n = fmt.Sprintf("%d", opts.BlockNumber.Int64())
+		}
+
+		return nil, errors.Wrapf(err, "failed to get contract state at block number '%s'", n)
+	}
+
+	ids := make([]*big.Int, 0, int(state.State.NumUpkeeps.Int64()))
+	for int64(len(ids)) < state.State.NumUpkeeps.Int64() {
+		startIndex := int64(len(ids))
+		maxCount := state.State.NumUpkeeps.Int64() - startIndex
+
+		if maxCount == 0 {
+			break
+		}
+
+		if maxCount > ActiveUpkeepIDBatchSize {
+			maxCount = ActiveUpkeepIDBatchSize
+		}
+
+		batchIDs, err := r.registry.GetActiveUpkeepIDs(opts, big.NewInt(startIndex), big.NewInt(maxCount))
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to get active upkeep IDs from index %d to %d (both inclusive)", startIndex, startIndex+maxCount-1)
+		}
+
+		ids = append(ids, batchIDs...)
+	}
+
+	return ids, nil
+}
+
+func (r *EvmRegistry) doCheck(ctx context.Context, keys []types.UpkeepKey, chResult chan checkResult) {
+	upkeepResults, err := r.checkUpkeeps(ctx, keys)
+	if err != nil {
+		chResult <- checkResult{
+			err: err,
+		}
+		return
+	}
+
+	upkeepResults, err = r.simulatePerformUpkeeps(ctx, upkeepResults)
+	if err != nil {
+		chResult <- checkResult{
+			err: err,
+		}
+		return
+	}
+
+	for i, res := range upkeepResults {
+		_, id, err := blockAndIdFromKey(res.Key)
+		if err != nil {
+			continue
+		}
+
+		up, ok := r.active[id.Int64()]
+		if ok {
+			upkeepResults[i].ExecuteGas = up.PerformGasLimit
+		}
+	}
+
+	chResult <- checkResult{
+		ur: upkeepResults,
+	}
+}
+
+func (r *EvmRegistry) checkUpkeeps(ctx context.Context, keys []types.UpkeepKey) ([]types.UpkeepResult, error) {
+	var (
+		checkReqs    = make([]rpc.BatchElem, len(keys))
+		checkResults = make([]*string, len(keys))
+	)
+
+	for i, key := range keys {
+		block, upkeepId, err := blockAndIdFromKey(key)
+		if err != nil {
+			return nil, err
+		}
+
+		opts, err := r.buildCallOpts(ctx, block)
+		if err != nil {
+			return nil, err
+		}
+
+		blockRef := "latest"
+		if opts.BlockNumber != nil {
+			blockRef = hexutil.EncodeBig(opts.BlockNumber)
+		}
+
+		payload, err := r.abi.Pack("checkUpkeep", upkeepId)
+		if err != nil {
+			return nil, err
+		}
+
+		var result string
+		checkReqs[i] = rpc.BatchElem{
+			Method: "eth_call",
+			Args: []interface{}{
+				map[string]interface{}{
+					"to":   r.addr.Hex(),
+					"data": hexutil.Bytes(payload),
+				},
+				blockRef,
+			},
+			Result: &result,
+		}
+
+		checkResults[i] = &result
+	}
+
+	if err := r.client.BatchCallContext(ctx, checkReqs); err != nil {
+		return nil, err
+	}
+
+	var (
+		err     error
+		results = make([]types.UpkeepResult, len(keys))
+	)
+
+	for i, req := range checkReqs {
+		if req.Error != nil {
+			if strings.Contains(req.Error.Error(), "reverted") {
+				// subscription was canceled
+				// NOTE: would we want to publish the fact that it is inactive?
```

Comment:
Not sure when this will happen. In 2.0 checkUpkeep should never revert, instead return false with error code